sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,558 @@
+		// Wait for no other exclusive lock.
+		mutex.waitFor(noWriteLock);
+		count += Exclusive;
+		mutex.waitFor(hasExclusiveLock);
+	}
+
+	void exclusiveUnlockImpl() {
+		assert(mutex.isHeld());
+		assert((count & ~Exclusive) > 0);
+		--count;
+	}
+
+	void unlockWriteImpl() {
+		assert(mutex.isHeld());
+		assert(count == Exclusive);
+		count = 0;
+	}
+}
+
+unittest sharedLocks {
+	// Simple testing of shared locks, then test multiple threads acquiring
+	// the lock.
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+
+	lock.sharedLock();
+	assert(lock.count == 1);
+	assert(!lock.mutex.isHeld());
+
+	lock.sharedLock();
+	assert(lock.count == 2);
+
+	lock.sharedUnlock();
+	assert(lock.count == 1);
+
+	import d.sync.atomic;
+	shared Mutex mutex;
+	shared Atomic!uint numLocked;
+	bool exitThread;
+
+	void* run() {
+		lock.sharedLock();
+
+		numLocked.fetchAdd(1);
+
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		bool shouldExit() {
+			return exitThread;
+		}
+
+		mutex.waitFor(shouldExit);
+		lock.sharedUnlock();
+		numLocked.fetchSub(1);
+		return null;
+	}
+
+	import core.stdc.pthread;
+	pthread_t[50] tids;
+	foreach (i; 0 .. tids.length) {
+		tids[i] = runThread(run);
+	}
+
+	{
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		bool allThreadsReady() {
+			return numLocked.load() == tids.length;
+		}
+
+		mutex.waitFor(allThreadsReady);
+		assert(lock.count == tids.length + 1);
+
+		exitThread = true;
+
+		bool allThreadsDone() {
+			return numLocked.load() == 0;
+		}
+
+		mutex.waitFor(allThreadsDone);
+		assert(lock.count == 1);
+	}
+
+	void* ret;
+	foreach (i; 0 .. tids.length) {
+		pthread_join(tids[i], &ret);
+	}
+
+	lock.sharedUnlock();
+	assert(lock.count == 0);
+}
+
+unittest exclusiveLock {
+	// Test using the exclusive lock. Ensure only one thread can hold the
+	// lock at any given time.
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+
+	lock.exclusiveLock();
+	assert(lock.count == SharedLock.Exclusive);
+
+	lock.exclusiveUnlock();
+	assert(lock.count == 0);
+
+	import d.sync.atomic;
+	shared Atomic!uint state;
+
+	void* run1() {
+		assert(state.load() == 0);
+
+		lock.exclusiveLock();
+		assert(lock.count == SharedLock.Exclusive, "Invalid lock state!");
+		assert(lock.mutex.isHeld(), "Mutex not held while exclusively locked!");
+		state.store(1);
+		lock.exclusiveUnlock();
+
+		while (state.load() != 2) {
```

Comment:
Why not a mutex/waitFor here?

Reply:
I see, because it is what the mutex test does :)

Note that the mutex test predates waitFor, which would be a better tool here.

Reply:
Fixed.